snappy/DiskCache_DiskCache.py

```diff
@@ -250,46 +251,80 @@ def getSymbolFileRelPath(self, libName, breakpadId, fileName):
         return path
 
     def downloadToCache(self, libName, breakpadId, symbolFilename, destPath, saveRaw=False):
-        destDir = os.path.dirname(destPath)
-        if not os.path.exists(destDir):
-            os.makedirs(destDir)
-
-        for symbolURL in self.symbolURLs:
-            url = self.getSymbolURL(symbolURL, libName, breakpadId, symbolFilename)
-            try:
-                with contextlib.closing(urllib2.urlopen(url)) as response:
-                    if response.getcode() != 200:
-                        logger.log(logLevel.WARNING,
-                                   "Got HTTP Code {} when requesting symbol file at {}"
-                                   .format(response.getcode(), url))
-                        continue
-                    headers = response.info()
-                    contentEncoding = headers.get("Content-Encoding", "").lower()
-                    if contentEncoding in ("gzip", "x-gzip", "deflate"):
-                        with contextlib.closing(StringIO(response.read())) as dataStream:
-                            try:
-                                with gzip.GzipFile(fileobj=dataStream) as f:
-                                    data = f.read()
-                            except Exception:
-                                data = dataStream.decode('zlib')
-                    else:
-                        data = response.read()
-                libId = "{}/{}/{}".format(libName, breakpadId, symbolFilename)
-                with open(destPath, 'wb') as fp:
-                    if saveRaw:
+        urlsToRetry = self.symbolURLs
+        for attempt in xrange(config['retries']):
+            symbolURLs = urlsToRetry
+            urlsToRetry = []
+            for symbolURL in symbolURLs:
+                url = self.getSymbolURL(symbolURL, libName, breakpadId, symbolFilename)
+                success, exists, data = self.retrieveFile(url)
+                if not success:
+                    urlsToRetry.append(symbolURL)
+                    continue
+                if not exists:
+                    # Don't retry this server if we know the file is not on it
+                    continue
```

Comment:
I would change `retrieveFile` such that it took all the urls and did all the work to retrieve the file from whichever url worked and then returned the file or raised an error that it wasn't retrievable. Everything above this point in `downloadToCache` should be in that. So then it's something like:

```
def downloadToCache(self, libName, breakpadId, symbolFilename, destPath, saveRaw=False):
    try:
        data = self.retrieveFile(self.symbolUrls, libName, breakpadId, symbol_filename, num_retries=config['retries'])
    except NoFileError:
        return False

    if not saveRaw:
    ....

def retreiveFile(self, urls, libname, breakpad_id, symbol_filename, num_retries):
    skip_urls = []
    for attempt in range(num_retries):
        for url in urls:
            if url in skip_urls:
                continue
            symbol_url = self.getSymbolURL(url, libname, breakpad_id, symbol_filename)
        ...
```

Another thing you want to think about is that retrying immediately usually isn't helpful for ephemeral errors which take some time to go away, so you want to wait a bit between retries.

Sometimes you can do that with `time.sleep`. With Antenna, I'm going to use `gevent.sleep` because I'm using gevent. I have no idea what the equivalent is for Tornado apps.